src/main/java/org/folio/circulation/domain/validation/LoanAndRelatedRecordsValidator.java

```diff
@@ -0,0 +1,64 @@
+package org.folio.circulation.domain.validation;
+
+import static org.apache.commons.lang3.exception.ExceptionUtils.getStackTrace;
+import static org.folio.circulation.support.Result.failed;
+import static org.folio.circulation.support.Result.succeeded;
+import static org.folio.circulation.support.results.CommonFailures.failedDueToServerError;
+import static org.folio.circulation.support.ValidationErrorFailure.singleValidationError;
+
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.ValidationErrorFailure;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LoanAndRelatedRecordsValidator {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private LoanAndRelatedRecordsValidator() {}
+
+  public static Result<LoanAndRelatedRecords> refuseWhenLoanDueDateUpdateOnClaimedReturned(Result<LoanAndRelatedRecords> result, LoanRepository loanRepository) {
+    return result.failWhen(
+      r -> {
+        Result<Loan> loanResult;
+        CompletableFuture<Result<Loan>> future = loanRepository.getById(r.getLoan().getId());
+
+        try {
+          loanResult = future.get();
```

Comment:
This is a blocking operation, we have not to use such kind of operations in production code.

You can refactor it something like:
* `org.folio.circulation.resources.LoanCollectionResource#replace` change to:
```java
.thenCombineAsync(loanRepository.getById(loan.getId()),
        LoanAndRelatedRecordsValidator::refuseWhenLoanDueDateUpdateOnClaimedReturned)
```
* Then the `LoanAndRelatedRecordsValidator`:
```java
public static Result<LoanAndRelatedRecords> refuseWhenLoanDueDateUpdateOnClaimedReturned(
    Result<LoanAndRelatedRecords> result, Result<Loan> oldLoanResult) {

    return result.combineToResult(oldLoanResult,
      (ctx, oldLoan) -> {
        if (!Objects.equals(ctx.getLoan().getDueDate(), oldLoan.getDueDate())
          && ctx.getLoan().getItem().isClaimedReturned()) {
          return failed(dueDateChangedFailedForClaimedReturned(ctx));
        }

        return succeeded(ctx);
      }
    );
  }
```

Reply:
The `thenCombine()`/`thenCombineAsync()` will pass along the exceptions. Which is a problem.
I have already attempted this here: https://github.com/folio-org/mod-circulation/pull/480/commits/61d29106ca20b9dc5124704080119b6c3380d9ce

After some discussion, it was determined that I would try the `thenApply()` approach (in part due to time constraints).

I believe that blocking operations, such as `thenApply()` or `thenCombine()` would be necessary to ensure that the existence (or lack thereof) of a duplicate dueDate can be confirmed. Maybe I am overthinking this and it doesn't need the blocking because there is a blocking operation later on before the final write operation that would be affected by the result.

I will attempt your particular approach to see if it is any more successful than the one I tried in the provided commit.

Reply:
Unfortunately, it looks like when using your `thenCombineAsync()` example, the same problem that I described in the said commit still happens.

Look at the following test, for example:
`canCreateClosedLoanInSpecificLocationWithoutUserId()` in `LoanAPITests`.

Reply:
Yeah, I see your point now. Well, the test case is strange, why do we use put for record create?..

@marcjohnson-kint, do you have any idea? I think it should be fixed to use post.

Reply:
Seems it is related to https://issues.folio.org/browse/CIRC-136. 

Maybe we should adjust the test case and create a loan with user and then update it with no user. What do you think @marcjohnson-kint, @kaladay?

Reply:
@bohdan-suprun In general, `PUT` is intended to create records at specific resource addresses.

That said, there should be few tests in mod-circulation that actually do that.

It has highlighted that not all PUT requests should be rejected when the status is `Claimed returned` or `Declared lost`

Yes, anonymisation relies on a loan being created with a borrower and then eventually that borrower is removed after the loan has been closed.

Reply:
@bohdan-suprun I think we should do something other than changing the tests. That said, I would prefer to not use the current design here (that unideal try..catch..etc.. structure).

I worry about fixing the test case as doing this may work around the problem in the tests, but what does production do? I am, as a general rule, against changing tests for these reasons even if making such changes is very tempting.

Reply:
Yeah, agree, we should keep supporting creating using PUT, UI can use this approach somewhere. 

Generally, we have to omit `CompletableFuture.get/join` as they blocks execution thread. Any other operations are fine - e.g. `Result.value`, `Future.get`.

Reply:
I've been experimenting with expressing this behaviour differently. I've pushed a [branch](https://github.com/folio-org/mod-circulation/tree/CIRC-637-experiment) with some rough changes that might help avoid some of the challenges around this change.

It might also allow for the claimed returned and declared lost validation to be combined. And in both cases, still allow general updates of a loan.

What do folks think?

The difficulty of this change tells me that there likely needs to be deeper changes made to this area of the code. For example, the two distinct processes in this endpoint could be [split](https://issues.folio.org/browse/CIRC-650).

It might also be that there is a need for more sophisticated flow of control with Completable Future and Result. This can be challenging to do nicely in Java due to how generic type erasure does not allow similar, yet separate overloads of methods.

Reply:
I reviewed and accepted @marcjohnson-kint  changes from the specified [branch](https://github.com/folio-org/mod-circulation/tree/CIRC-637-experiment).

I think that moving this forward in this way will allow for easier intergration with the declared lost validation (and subsequently the future planned changes for separating concerns with the due date).